gorm-tools/src/main/groovy/gorm/tools/repository/BulkableRepo.groovy

```diff
@@ -0,0 +1,162 @@
+    int batchSize  //XXX https://github.com/9ci/domain9/issues/331  test if @Value works on trait
+
+    @Value('${nine.autocash.parallelProcessing.enabled:false}')
+    boolean parallelProcessingEnabled //XXX https://github.com/9ci/domain9/issues/331  we might have to move it, here we can have default
+
+    // GormRepo implements it
+    abstract void bindAndCreate(D entity, Map data, Map args)
+
+    // need cleaner way to do transaction, change it
+    abstract gormStaticApi()
+    abstract  Class<D> getEntityClass()
+
+    //// XXX https://github.com/9ci/domain9/issues/331 Not sure if needed any more
+    //abstract List bulkCreate()
+
+//////// MOVED FROM GORM REPO
+
+    /**
+     * allows to pass in bulk of records at once, for example /api/book/bulk
+     * Each call creates a job that stores info for the call and is returned with results
+     * @param dataList the list of data maps to create
+     * @param args args to pass to doCreate. It can have:
+     *      onError // 'rollback' or 'continue' (catches error) or 'skip'
+     *      errorThreshold // number of errors to stop and rollback
+     *      jdbc.batch_size // all, default or specify number
+     *      gpars.poolsize
+     * @return Job
+     */
+    J bulkCreate(List<Map> dataList, Map args = [:]){
+
+        // create job
+        Class<J> jobClass = (Class<J>)GenericTypeResolver.resolveTypeArguments(getClass(), BulkableRepo)[1]
+        J job = jobClass.newInstance() as J
+
+        // XXX We can do something similar for error count
+        AtomicInteger count = new AtomicInteger(0)
+
+        // for error handling -- based on `onError` from args commit success and report the failure or fail them all
+        //  also use errorThreshold, for example if it failed on more than 10 records we stop and rollback everything
+        //@jdabal - for parallel async, we can not rollback batches which are already commited
+        //as each batch would be in its own transaction.
+        List bulkResult = []
+        if(parallelProcessingEnabled) {
+            //run the batch parallel in batched transactions
+            asyncSupport.parallel([batchSize:batchSize], dataList.collate(batchSize)) { List<Map> batch ->
+                // returns entity that was created, but into Job object we only want id and sourceid.
+                // id:123, source.sourceId, -  @jdabal assume that every domain which is bulk imported haas sourceId field ?
+                // store results in resultList,but id and sourceId only for succesfully created records
+                List results = doBulkCreate(batch, args)
+                if(results) bulkResult.addAll transformResults(results)
+            }
+        } else {
+            List results = doBulkCreate(dataList, args)
+            if(results) bulkResult.addAll transformResults(results)
+        }
+
+        count.getAndAdd(bulkResult.size())
+
+        job['results'] = Jsonify.render(bulkResult).jsonText.bytes
+        job.persist()
```

Comment:
shouldn't we be doing this stuff in a try/finally so we are sure its done? like spring and grails does for transactions wrappers?